File: rb101/rb101-rb109-small_problems/easy2/sum_or_product.rb
def retrieve_number
  puts '>> Please enter an integer greater than 0:'
  gets.to_i
end

def valid_number?(number)
  number > 0 && number.is_a?(Integer)
end

def display_invalid_input
  puts 'Oops! Invalid input!'
end

def retrieve_operation
  puts ">> Enter 's' to compute the sum, 'p' to compute the product."
  gets.chomp.downcase
end

def compute_sum(number)
  (1..number).inject(:+)
end

def compute_product(number)
  (1..number).inject(:*)
end

def display_sum(number, sum)
  puts "The sum of the integers between 1 and #{number} is #{sum}."
end

def display_product(number, product)
  puts "The product of the integers between 1 and #{number} is #{product}."
end

number = nil
loop do
  number = retrieve_number

  break if valid_number?(number)
  display_invalid_input
end

loop do
  operation = retrieve_operation

  case operation
  when 's'
    sum = compute_sum(number)
    display_sum(number, sum)
    break
  when 'p'
    product = compute_product(number)
    display_product(number, product)
    break
  else
    display_invalid_input
  end
end

# ALTERNATIVE SOLUTION:

# def compute_sum(number)
#   total = 0
#   1.upto(number) { |value| total += value }
#   total
# end

# def compute_product(number)
#   total = 1
#   1.upto(number) { |value| total *= value }
#   total
# end

File: js210/js210-small_problems/easy1/sumOrProduct.js
function computeSum(number) {
  let total = 0;

  for (let currentNumber = 1; currentNumber <= number; currentNumber += 1) {
    total += currentNumber;
  }

  return total;
}

function computeProduct(number) {
  let total = 1;

  for (let currentNumber = 1; currentNumber <= number; currentNumber += 1) {
    total *= currentNumber;
  }

  return total;
}

const rlSync = require('readline-sync');

let number = parseInt(rlSync.question('Please enter an integer greater than 0: ', {
  limit: [/^[\d]+$/],
  limitMessage: 'Oops! Invalid number!'
}), 10);

let operation = rlSync.question('Enter "s" to compute the sum, or "p" to compute the product: ', {
  limit: ['p', 's'],
  limitMessage: 'Oops! Unknown operation! Please enter "s" or "p"!'
}).toLowerCase();

if (operation === 's') {
  let sum = computeSum(number);
  console.log(`The sum of the integers between 1 and ${String(number)} is ${sum}.`);
} else {
  let product = computeProduct(number);
  console.log(`The product of the integers between 1 and ${String(number)} is ${product}.`);
}

// FURTHER EXPLORATION:

// function computeSum(numbers) {
//   return numbers.reduce((acc, num) => acc + num);
// }

// function computeProduct(numbers) {
//   return numbers.reduce((acc, num) => acc * num);
// }

// console.log(computeSum([1, 2, 3, 4, 5]));        // 15
// console.log(computeProduct([1, 2, 3, 4, 5, 6])); // 720
